### ISSUE TITLE
Android app for client

### ISSUE THREAD

**Issue by ingnelson** [bernardkkt/netwithquic#1]
Hi.
Can this works like a VPN or bypass a firewall?
If the answer is yes, can you create a Android app for client ?

Thanks.

**Comment by bernardkkt** [comment]
Hi @ingnelson 

The tool I am using for the client and the server is Project V (aka v2ray), and I would say it works as a proxy service rather than VPN when **it runs on PCs**, as it does not come with a tun interface to capture **ALL** network packets and send them to a designated server. 

However, it is possible to run it like a VPN on Android. There are a few Project V clients available on Google Play and GitHub, and probably one that I could recommend you is [Kitsunebi by eycorsican](https://github.com/eycorsican/kitsunebi-android). In that case, after generating your own configuration on https://bernardkkt.github.io/netwithquic/, you will need to enter the configuration manually on the Android client if the client could not read the configuration file directly.

**Comment by ingnelson** [comment]
@bernardkkt can you tell me what I am doing wrong to run the server , please .
On UBUNTU VPS 
1) I installed docker : sudo curl -sS https://get.docker.com/ | sh
2) docker build -t "netwithquic" https://github.com/bernardkkt/netwithquic.git
3)run docker:   docker run -v -p 0.0.0.0:80:80 /etc/v2ray/config.json netwithquic

Error:
docker: invalid reference format.
See 'docker run --help'.

Can you tell me what command do I need to write or put in , please .

**Comment by bernardkkt** [comment]
@ingnelson I believe I did a small mistake in my guide :joy: . Thanks for spotting the error.

Now instead of doing this: 
> docker run -v -p 0.0.0.0:80:80 /etc/v2ray/config.json netwithquic

Change it to:
```
docker run -v $PWD:/etc/v2ray -p 0.0.0.0:$PORT:$PORT netwithquic
```

Let me know if it helps. :smile:

**Comment by ingnelson** [comment]
@bernardkkt I did like you said  and the server is running , but I do not know if it is ok liike this : 
docker run -v /etc/v2ray/config.json -p 0.0.0.0:80:80 netwithquic

One question : do I have to put twice port (example: 0.0.0.0:80:80 ) ?
there are 2 errors :
1) when I import the client configuration from clipboard to kitsunebi app and then I try to run the app , the app said : "invalid V2RAY config".

2) if I do it manually , the app does not connect to the server or not create the tunnel using kitsunebi app or V2RAY app

Client : 
{"inbounds":[{"tag":"socks-in","port":1080,"listen":"::","protocol":"socks","settings":{"udp":true}}],"outbounds":[{"protocol":"vmess","settings":{"vnext":[{"address":"172.245.22.211","port":"80","users":[{"id":"c580d7ce-4050-47af-af29-7ef05b3d26e6","alterId":"4"}]}]},"streamSettings":{"network":"quic","quicSettings":{"security":"chacha20-poly1305","key":"test123","header":{"type":"none"}}}}]}


Server:

{"inbounds":[{"port":"80","protocol":"vmess","settings":{"clients":[{"id":"c580d7ce-4050-47af-af29-7ef05b3d26e6","alterId":"4"}]},"streamSettings":{"network":"quic","quicSettings":{"security":"chacha20-poly1305","key":"test123","header":{"type":"none"}}}}],"outbounds":[{"protocol":"freedom"}]}

What can be wrong ?, help

**Comment by bernardkkt** [comment]
@ingnelson If you ran the command with the correct `-v` parameters, i.e. mapping the path of the folder that contains the configuration file to `/etc/v2ray`, then you are doing it fine at the server side.

Yes on the server side you need to put the port twice. That means Docker will try to bind the host and the container via port 80 (port 80 -> port 80). In order to make it more specific (if it isn't necessary), change `-p 0.0.0.0:80:80` to `-p 0.0.0.0:80:80/udp` so that the QUIC packets can be reached to the container. Your last resort will be replacing `-p 0.0.0.0:80:80/udp` to `--net=host` if it still doesn't work.

Your configuration looks fine, but I noticed that the generated configuration is using string type for `port` and `alterId` instead of int type. I have done some fixes and please generate the configurations again. Make sure you have entered the IP address of your server correctly for your client configuration.

Side note: you might want to remove sensitive information regarding your configuration from the comment.

**Comment by ingnelson** [comment]
@bernardkkt I did like you said , but it does not open any websites on Android .
Have you try it , before ?

By the way can you create a proxy that only send a URI and Begin and finish with a "#" .
The proxy has to send on the URI the password encode maybe base64 , similar to this : 

GET #verx=2&0=c610d32167c94e46a48d959354202365&1=34c961c67ff3b923603867b1aecd506b&2=0&3=1&4=4&ope=1&cy=Claro-ni&7=dx1iv8o-nX8xY4AUPc9tWA==# HTTP/1.1

![IMG_20200501_212110](https://user-images.githubusercontent.com/6626845/83301991-cbe39c00-a1b7-11ea-8c9d-78702ee15db4.jpg)
![IMG_20200429_101602](https://user-images.githubusercontent.com/6626845/83302003-cf772300-a1b7-11ea-877a-39f11381d991.jpg)



Like you see there no http headers like : host , connect , upgrade , origin etc. Only the URI.

Please , because with that fragment http request  proxy + shadowsocks or V2RAY it is the only way to bypass my ISP firewall to get internet.

You can create it with python or other languages.

Here is the packet data , you can see with Wireshark program on PC or online on :
https://www.cloudshark.org/login
User: qojmndw782@aprimail.com
Pass: test123

Note: I just uploaded the file to www.cloudshark.org/login

[quantumnica1.zip](https://github.com/bernardkkt/netwithquic/files/4704126/quantumnica1.zip)

**Comment by bernardkkt** [comment]
@ingnelson I am afraid I could only help you on how to make netwithquic work at your side. I will try my best.

Yes I did try it and it worked. In fact I was doing the test locally with your configuration.
Below is the server configuration I have used in my testing:
```
{"inbounds":[{"port":80,"protocol":"vmess","settings":{"clients":[{"id":"c580d7ce-4050-47af-af29-7ef05b3d26e6","alterId":4}]},"streamSettings":{"network":"quic","quicSettings":{"security":"chacha20-poly1305","key":"test123","header":{"type":"none"}}}}],"outbounds":[{"protocol":"freedom"}]}
```
Take note with the values for **port** and **alterId**: they are integer, not string.

I ran the server command on my machine with the configuration above:
```
$ sudo docker run -v $PWD:/etc/v2ray -p 0.0.0.0:80:80/udp netwithquic
```
This is what I see right after running the command:
```
2020/05/30 02:04:39 Using default config:  /etc/v2ray/config.json
V2Ray 4.23.1 (V2Fly, a community-driven edition of V2Ray.) Custom (go1.13 linux/amd64)
A unified platform for anti-censorship.
2020/05/30 02:04:39 [Info] v2ray.com/core/common/platform/ctlcmd: <v2ctl message> 
v2ctl> Read config:  /etc/v2ray/config.json
2020/05/30 02:04:39 [Warning] v2ray.com/core: V2Ray 4.23.1 started
```
Then I opened Kitsunebi on my Android phone, pressed on the **+** button at the top right corner, chose **Add Endpoint**, then chose **Manual**.
I filled in the form as below:
![86EmhSzT](https://user-images.githubusercontent.com/16949329/83317634-a2327f80-a260-11ea-9931-3cfb6b474e0a.jpg)
Saved it, chose the profile and pressed the play button at the bottom right corner. 
Now I have all my network traffic routed to my machine.

If you find out that it works when you run the command at the server but not right after you close your server console session, then you might need to run Docker in detach mode. Simply add the parameter `-d` right after `docker run`.

Hope it helps.
